internal/controller/apiextensions/claim/syncer_ssa.go

```diff
@@ -0,0 +1,292 @@
+	//
+	// 1. We remove CSA manager "crossplane", triggering XR controller watch
+	// 2. XR controller uses CSA manager "crossplane", triggering our watch
+	// 3. Back to step 1 :)
+	//
+	// In practice we only need to upgrade once, to ensure we don't share fields
+	// that only this controller has ever applied with "crossplane". We assume
+	// that if our SSA manager already exists, we've done the upgrade.
+	for _, e := range obj.GetManagedFields() {
+		if e.Manager == ssaManager {
+			return nil
+		}
+	}
+	p, err := csaupgrade.UpgradeManagedFieldsPatch(obj, sets.New[string](csaManagers...), ssaManager)
+	if err != nil {
+		return errors.Wrap(err, errCreatePatch)
+	}
+	if p == nil {
+		// No patch means there's nothing to upgrade.
+		return nil
+	}
+	return errors.Wrap(resource.IgnoreNotFound(u.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), errPatchFieldManagers)
+}
+
+// A ServerSideCompositeSyncer binds and syncs a claim with a composite resource
+// (XR). It uses server-side apply to update the XR.
+type ServerSideCompositeSyncer struct {
+	client client.Client
+	names  names.NameGenerator
+}
+
+// NewServerSideCompositeSyncer returns a CompositeSyncer that uses server-side
+// apply to sync a claim with a composite resource.
+func NewServerSideCompositeSyncer(c client.Client, ng names.NameGenerator) *ServerSideCompositeSyncer {
+	return &ServerSideCompositeSyncer{client: c, names: ng}
+}
+
+// Sync the supplied claim with the supplied composite resource (XR). Syncing
+// may involve creating and binding the XR.
+func (s *ServerSideCompositeSyncer) Sync(ctx context.Context, cm *claim.Unstructured, xr *composite.Unstructured) error { //nolint:gocyclo // This complex process seems easier to follow in one long method.
+	// First we sync claim -> XR.
+
+	// Create an empty XR patch object. We'll use this object to ensure we only
+	// SSA our desired state, not the state we previously read from the API
+	// server.
+	xrPatch := composite.New(composite.WithGroupVersionKind(xr.GroupVersionKind()))
+
+	// If the claim references an XR, make sure we're going to apply that XR. We
+	// do this instead of using the supplied XR's name just in case the XR
+	// exists, but we couldn't get it due to a stale cache.
+	if ref := cm.GetResourceReference(); ref != nil {
+		xrPatch.SetName(ref.Name)
+	}
+
+	// If the XR doesn't have a name (i.e. doesn't exist), derive a name from
+	// the claim. The generated name is likely (but not guaranteed) to be
+	// available when we create the XR. If taken, then we are going to update an
+	// existing XR, probably hijacking it from another claim.
+	if xrPatch.GetName() == "" {
+		xrPatch.SetGenerateName(fmt.Sprintf("%s-", cm.GetName()))
+		if err := s.names.GenerateName(ctx, xrPatch); err != nil {
+			return errors.Wrap(err, errGenerateName)
+		}
+	}
+
+	// It's possible we're being asked to configure a statically provisioned XR.
+	// We should respect its existing external name annotation.
```

Comment:
Looks like there was a Non-Breaking Space:
```suggestion
	// We should respect its existing external name annotation.
```

Reply:
Weird, not sure how that got there.